src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Connection.cs

```diff
@@ -622,80 +631,125 @@ private void ProcessGoAwayFrame(FrameHeader frameHeader)
             _incomingBuffer.Discard(frameHeader.Length);
         }
 
-        private async ValueTask AcquireWriteLockAsync()
+        private async ValueTask StartWriteAsync(int writeBytes, CancellationToken cancellationToken = default, bool acquireWriteLock = true)
         {
-            await _writerLock.WaitAsync().ConfigureAwait(false);
-
-            // If the connection has been aborted, then fail now instead of trying to send more data.
-            if (IsAborted())
+            if (acquireWriteLock)
             {
-                throw new IOException(SR.net_http_invalid_response);
+                await AcquireWriteLockAsync(cancellationToken).ConfigureAwait(false);
             }
-        }
 
-        private void ReleaseWriteLock()
-        {
-            // Currently, we always flush the write buffer before releasing the lock.
-            // If we change this in the future, we will need to revisit this assert.
-            Debug.Assert(_outgoingBuffer.ActiveMemory.IsEmpty);
+            try
+            {
+                // If there is a pending write that was canceled while in progress, wait for it to complete.
+                if (_inProgressWrite != null)
+                {
+                    await _inProgressWrite;
+                    _inProgressWrite = null;
+                }
+
+                int totalBufferLength = _outgoingBuffer.TotalLength;
+                int activeBufferLength = _outgoingBuffer.ActiveSpan.Length;
 
-            _writerLock.Release();
+                if (totalBufferLength >= 32768 &&
+                    writeBytes >= totalBufferLength - activeBufferLength &&
+                    activeBufferLength > 0)
+                {
+                    // If the buffer has already grown to 32k, does not have room for the next request,
+                    // and is non-empty, flush the current contents to the wire.
+                    await FlushOutgoingBytesAsync().ConfigureAwait(false);
+                }
+                _outgoingBuffer.EnsureAvailableSpace(writeBytes);
+            }
+            catch
+            {
+                if (acquireWriteLock)
+                {
+                    _writerLock.Release();
+                }
+                throw;
+            }
         }
 
-        private async ValueTask SendSettingsAckAsync()
+        // This method handles flushing bytes to the wire. Writes here need to be atomic, so as to avoid
+        // killing the whole connection. Callers must hold the write lock, but can specify whether or not
```

Comment:
This won't validate that the actual caller holds the write lock, but you could at least do `Debug.Assert(_writerLock.CurrentCount == 0)` to check that _someone_ holds the write lock.

Reply:
Sounds reasonable. I'll add this.